examples/Multitest/Parallel/parallel_tasks.py

```diff
@@ -19,46 +22,53 @@ class SampleTest(object):
     """
 
     def __init__(self):
-        self._test_g2_1_done = False
+        self._test_g2_1_done = threading.Event()
+        self._barrier = thread_utils.Barrier(2)
 
     @testcase(execution_group='first')
     def test_g1_1(self, env, result):
-        """Wait for test_g1_2 to also acquire the first resource."""
-        with env.resources['first'] as res:
-            result.true(res.active)
-            while res.refcount < 2:
-                result.log('Waiting for test_g1_2...')
-                time.sleep(1)
-
-    @testcase(execution_group='second')
```

Comment:
From the look of the original example, testplan logic is to sort the testcases by group before execution. I think this behavior is very implicit, we should not do that. All testcases should be executed in the definition order, then adjacent testcases of the same execution_group can be run in parallel.

Reply:
As discussed, the behaviour to run testcases in the order of definition applies only when running a MultiTest serially. When run in parallel, tests are run in order of their groups, and ordering within a group is undefined since the tests are run in parallel.

I will move back the ordering of testcases and add a comment to the example to describe this behaviour.

Reply:
In any case, changing the behaviour of parallel MTests is not in scope of this PR and would require more careful thought/design. I'm only trying to fix the example to be less flaky so it doesn't break our CI, as well as to follow best practices for multithreading.